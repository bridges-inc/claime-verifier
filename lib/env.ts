const projectName = 'claime-verifier'

export enum Environments {
  PROD = 'prod',
  DEV = 'dev',
  TEST = 'test',
}
export interface EnvironmentVariables {
  allowedOrigin: string
  rootDomain: string
  hostedZoneId: string
  certificateArn: string
}

const EnvironmentVariablesSetting: {
  [key in Environments]: EnvironmentVariables
} = {
  [Environments.PROD]: {
    allowedOrigin: '',
    rootDomain: '',
    hostedZoneId: '',
    certificateArn: '',
  },
  [Environments.DEV]: {
    allowedOrigin: 'https://claime-webfront-k6p1srx99-squard.vercel.app',
    rootDomain: 'claime-dev.tk',
    hostedZoneId: 'Z08620181ARYV5PENJUEI',
    certificateArn:
      'arn:aws:acm:us-east-1:495476032358:certificate/92761d01-6e38-4ca4-8580-a945a5c379fe',
  },
  [Environments.TEST]: {
    allowedOrigin: '',
    rootDomain: 'test',
    hostedZoneId: 'test',
    certificateArn: '',
  },
}

export function valueOf(env: Environments): EnvironmentVariables {
  return EnvironmentVariablesSetting[env]
}
export const withEnvPrefix = (target: Environments, str: string) =>
  `${projectName}-${str}-${target}`

export const isProd = (target: Environments) => {
  return target === Environments.PROD
}
export const isDev = (target: Environments) => {
  return target === Environments.DEV
}
