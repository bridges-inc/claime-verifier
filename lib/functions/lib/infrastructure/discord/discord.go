package discord

import (
	"bytes"
	"claime-verifier/lib/functions/lib/common/log"
	"claime-verifier/lib/functions/lib/guild"
	"crypto/ed25519"
	"encoding/hex"
	"encoding/json"
	"errors"
	"net/http"
	"strconv"

	"github.com/bwmarrin/discordgo"
)

type (
	PubKeyResolver interface {
		DiscordPublicKey() (val string, err error)
	}
	KeyResolver interface {
		PubKeyResolver
		DiscordBotToken() (val string, err error)
		ClaimePrivateKey() (val ed25519.PrivateKey, err error)
	}
	InteractionConverter struct {
		pubkey    string
		bottoken  string
		claimeKey ed25519.PrivateKey
	}
	InteractionResponse struct {
		Type int `json:"type"`
	}
)

func (i InteractionResponse) ShouldProcess() bool {
	return i.Type == 5
}

func HandleInteractionResponse(request map[string]interface{}) (InteractionResponse, error) {
	req, err := json.Marshal(request["jsonBody"])
	if err != nil {
		return InteractionResponse{}, err
	}
	webhook := discordgo.Webhook{}
	err = json.Unmarshal(req, &webhook)
	if err != nil {
		return InteractionResponse{}, err
	}
	if webhook.Type == discordgo.WebhookTypeChannelFollower {
		return InteractionResponse{Type: 5}, nil
	}
	if webhook.Type == discordgo.WebhookTypeIncoming {
		return InteractionResponse{Type: 1}, nil
	}
	return InteractionResponse{}, errors.New("unknown type:" + strconv.Itoa(int(webhook.Type)))
}

func ToInteraction(request map[string]interface{}) (discordgo.Interaction, error) {
	req, err := json.Marshal(request["jsonBody"])
	if err != nil {
		log.Error("", err)
		return discordgo.Interaction{}, err
	}
	interaction := discordgo.Interaction{}
	if err = interaction.UnmarshalJSON(req); err != nil {
		log.Error("", err)
		return discordgo.Interaction{}, err
	}
	return interaction, nil
}

// VerifyInteractionRequest verify signature of interaction request
func VerifyInteractionRequest(request map[string]interface{}, r PubKeyResolver) bool {
	pubkey, err := r.DiscordPublicKey()
	if err != nil {
		log.Error("", err)
		return false
	}
	return verify(request, pubkey)
}

func verify(request map[string]interface{}, publicKey string) bool {
	signature, ok := request["signature"].(string)
	if !ok {
		log.Info("sig not found")
		return false
	}
	timestamp, ok := request["timestamp"].(string)
	if !ok {
		log.Info("timestamp not found")
		return false
	}
	req, err := json.Marshal(request["jsonBody"])
	if err != nil {
		log.Error("", err)
		return false
	}

	httpreq, err := http.NewRequest("", "", bytes.NewReader(req))
	if err != nil {
		log.Error("", err)
		return false
	}
	key, err := hex.DecodeString(publicKey)
	if err != nil {
		log.Error("", err)
		return false
	}
	httpreq.Header.Add("X-Signature-Ed25519", signature)
	httpreq.Header.Add("X-Signature-Timestamp", timestamp)
	return discordgo.VerifyInteraction(httpreq, key)
}

func ToRegisterContractInput(d discordgo.ApplicationCommandInteractionData, guildID string) guild.ContractInfo {
	if len(d.Options) < 3 {
		return guild.ContractInfo{}
	}
	return guild.ContractInfo{
		RoleID:          d.Options[0].Value.(string),
		ContractAddress: d.Options[1].Value.(string),
		Network:         d.Options[2].Value.(string),
		GuildID:         guildID,
	}
}
