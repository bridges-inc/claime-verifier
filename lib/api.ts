import {
  DomainName,
  EndpointType,
  IResource,
  MockIntegration,
  PassthroughBehavior,
  RestApi,
  SecurityPolicy,
} from '@aws-cdk/aws-apigateway'
import * as certificatemanager from '@aws-cdk/aws-certificatemanager'
import { Code } from '@aws-cdk/aws-lambda'
import * as route53 from '@aws-cdk/aws-route53'
import * as alias from '@aws-cdk/aws-route53-targets'
import * as cdk from '@aws-cdk/core'
import { Stack } from '@aws-cdk/core'
import { resolve } from 'path'
import * as environment from './env'
import { hostedZoneFromId } from './route53'

type CordOptionsConfig = {
  methods: ('GET' | 'POST' | 'PUT' | 'DELETE')[]
  withCredentials?: true
  headers?: string[]
  origin?: string
}
export function addCorsOptions(
  apiResource: IResource,
  config: CordOptionsConfig,
) {
  const responseParameters = integrationResponseParameters(config)
  const methodResponseParameters = Object.keys(responseParameters).reduce(
    (res, key) => Object.assign(res, { [key]: true }),
    {},
  )
  apiResource.addMethod(
    'OPTIONS',
    new MockIntegration({
      integrationResponses: [{ statusCode: '200', responseParameters }],
      passthroughBehavior: PassthroughBehavior.NEVER,
      requestTemplates: { 'application/json': '{"statusCode": 200}' },
    }),
    {
      methodResponses: [
        { statusCode: '200', responseParameters: methodResponseParameters },
      ],
    },
  )
}

const integrationResponseParameters = (config: CordOptionsConfig) => {
  const { methods, withCredentials, headers, origin } = config
  const params = {
    'method.response.header.Access-Control-Allow-Headers': `'${
      headers?.join(',') || '*'
    }'`,
    'method.response.header.Access-Control-Allow-Origin': `'${origin || '*'}'`,
    'method.response.header.Access-Control-Allow-Methods': `'OPTIONS,${methods.join(
      ',',
    )}'`,
  }
  if (withCredentials) {
    Object.assign(params, {
      'method.response.header.Access-Control-Allow-Credentials': "'true'",
    })
  }
  return params
}

export const environmentVariables = (target: environment.Environments) => {
  return {
    AllowedOrigin: environment.valueOf(target).allowedOrigin,
    EnvironmentId: target,
    SubgraphEndpoint: environment.valueOf(target).subgraphEndpoint,
  }
}
const aRecord = (
  stack: Stack,
  target: environment.Environments,
  domainName: string,
  customDomain: DomainName,
) => {
  new route53.ARecord(stack, 'RestApiARecord', {
    zone: hostedZoneFromId(stack, target),
    recordName: domainName,
    target: route53.RecordTarget.fromAlias(
      new alias.ApiGatewayDomain(customDomain),
    ),
  })
}

export const code = (dirname: string) => {
  return Code.fromAsset(
    resolve(`${__dirname}/../`, 'lib', 'functions', dirname, 'bin', 'main.zip'),
  )
}
export const withCustomDomain = (
  stack: cdk.Stack,
  api: RestApi,
  domain: string,
  target: environment.Environments,
) => {
  const customDomain = api.addDomainName(
    environment.withEnvPrefix(target, 'domain'),
    customDomainProps(stack, domain, target),
  )
  aRecord(stack, target, domain, customDomain)
  return customDomain
}

const customDomainProps = (
  stack: cdk.Stack,
  domain: string,
  target: environment.Environments,
) => {
  return {
    domainName: domain,
    certificate: certificatemanager.Certificate.fromCertificateArn(
      stack,
      'Cert',
      environment.valueOf(target).certificateArn,
    ),
    securityPolicy: SecurityPolicy.TLS_1_2,
    endpointType: EndpointType.REGIONAL,
  }
}
