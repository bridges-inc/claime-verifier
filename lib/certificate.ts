import {
  DnsValidatedCertificate,
  ICertificate,
  ValidationMethod,
} from '@aws-cdk/aws-certificatemanager'
import { Construct, Stack } from '@aws-cdk/core'
import * as environment from './env'
import { hostedZoneFromId } from './route53'

export class CertificateStack extends Stack {
  public readonly certificate: ICertificate

  constructor(scope: Construct, id: string, target: environment.Environments) {
    super(scope, id)
    const { hostedZoneId } = environment.valueOf(target)
    if (environment.isProd(target)) return
    if (!hostedZoneId) throw new Error('env.hostedZoneId is requied')
    this.certificate = certificate(this, target)
  }
}

const certificate = (
  scope: Construct,
  target: environment.Environments,
): ICertificate => {
  const { rootDomain } = environment.valueOf(target)

  return new DnsValidatedCertificate(scope, 'Certificate', {
    domainName: `${rootDomain}`,
    subjectAlternativeNames: [`*.${rootDomain}`],
    hostedZone: hostedZoneFromId(scope, target),
    validationMethod: ValidationMethod.DNS,
    region: 'us-east-1',
  })
}
