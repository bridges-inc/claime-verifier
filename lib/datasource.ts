import {
  AttributeType,
  BillingMode,
  Table,
  TableProps,
} from "@aws-cdk/aws-dynamodb";
import * as cdk from "@aws-cdk/core";
import * as environment from "../lib/env";
export class DatasourceStack extends cdk.Stack {
  constructor(
    scope: cdk.Construct,
    id: string,
    target: environment.Environments,
    props?: cdk.StackProps
  ) {
    super(scope, id, props);
    const table = new Table(
      this,
      "claime-verifier-table",
      ddbProps(environment.withEnvPrefix(target, "main"))
    );
    table.addGlobalSecondaryIndex({
      indexName: "GSI-1",
      partitionKey: {
        name: "PK",
        type: AttributeType.STRING,
      },
      sortKey: {
        name: "Timestamp",
        type: AttributeType.STRING,
      },
    });
  }
}

const ddbProps = (tableName: string): TableProps => {
  return {
    tableName: tableName,
    partitionKey: {
      name: `PK`,
      type: AttributeType.STRING,
    },
    sortKey: {
      name: `SK`,
      type: AttributeType.STRING,
    },
    billingMode: BillingMode.PAY_PER_REQUEST,
    removalPolicy: cdk.RemovalPolicy.DESTROY,
    pointInTimeRecovery: true,
  };
};
