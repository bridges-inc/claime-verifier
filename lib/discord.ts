import {
  ApiKeySourceType,
  LambdaIntegration,
  RequestValidator,
  RestApi,
} from '@aws-cdk/aws-apigateway'
import { SubnetType, Vpc } from '@aws-cdk/aws-ec2'
import {
  Cluster,
  ContainerImage,
  FargateService,
  FargateTaskDefinition,
  LogDriver,
} from '@aws-cdk/aws-ecs'
import {
  Effect,
  ManagedPolicy,
  PolicyStatement,
  Role,
  ServicePrincipal,
} from '@aws-cdk/aws-iam'
import { Code, Function, Runtime, Tracing } from '@aws-cdk/aws-lambda'
import { LogGroup } from '@aws-cdk/aws-logs'
import { Secret } from '@aws-cdk/aws-secretsmanager'
import * as cdk from '@aws-cdk/core'
import { RemovalPolicy } from '@aws-cdk/core'
import { resolve } from 'path'
import { dataSourceReadWritePolicyStatement } from './datasource'
import * as environment from './env'
import { environmentVariables } from './restapi'

export class DiscordStack extends cdk.Stack {
  constructor(
    scope: cdk.Construct,
    id: string,
    target: environment.Environments,
    props?: cdk.StackProps,
  ) {
    super(scope, id, props)
    const api = new RestApi(this, 'RestApi', {
      restApiName: environment.withEnvPrefix(target, 'restapi'),
      apiKeySourceType: ApiKeySourceType.HEADER,
    })
    const discordAPISecrets = new Secret(
      this,
      environment.withEnvPrefix(target, 'discord-api-key'),
    )
    const func = discordFunction(
      this,
      this.region,
      this.account,
      'discord',
      target,
      api,
    )
    discordAPISecrets.grantRead(func)
    const vpc = new Vpc(this, 'vpc', {
      cidr: '192.156.0.0/16',
      natGateways: 1,
      subnetConfiguration: [
        {
          cidrMask: 24,
          name: 'Public1',
          subnetType: SubnetType.PUBLIC,
        },
        {
          cidrMask: 24,
          name: 'Public2',
          subnetType: SubnetType.PUBLIC,
        },
        {
          cidrMask: 24,
          name: 'Private1',
          subnetType: SubnetType.PRIVATE,
        },
        {
          cidrMask: 24,
          name: 'Private2',
          subnetType: SubnetType.PRIVATE,
        },
      ],
    })
    const executionRole = new Role(this, 'EcsTaskExecutionRole', {
      roleName: environment.withEnvPrefix(target, 'ecs-execution-role'),
      assumedBy: new ServicePrincipal('ecs-tasks.amazonaws.com'),
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName(
          'service-role/AmazonECSTaskExecutionRolePolicy',
        ),
      ],
    })
    const serviceTaskRole = new Role(this, 'EcsServiceTaskRole', {
      roleName: environment.withEnvPrefix(target, 'ecs-service-task-role'),
      assumedBy: new ServicePrincipal('ecs-tasks.amazonaws.com'),
    })
    basicPolicytStatements(this.region, this.account, target).forEach((s) =>
      serviceTaskRole.addToPolicy(s),
    )
    const logGroup = new LogGroup(this, 'ServiceLogGroup', {
      logGroupName: '/aws/ecs/claime-cluster' + target,
      removalPolicy: RemovalPolicy.DESTROY,
    })
    const image = ContainerImage.fromAsset('.')
    const cpu = 256
    const mem = 1024
    const taskDef = new FargateTaskDefinition(this, 'ServiceTaskDefinition', {
      cpu: cpu,
      memoryLimitMiB: mem,
      executionRole: executionRole,
      taskRole: serviceTaskRole,
    })
    taskDef.addContainer('ContainerDef', {
      image,
      cpu: cpu,
      memoryLimitMiB: mem,
      logging: LogDriver.awsLogs({
        streamPrefix: 'claime',
        logGroup,
      }),
      environment: environmentVariables(target),
    })
    const cluster = new Cluster(this, 'claime-cluster', {
      clusterName: environment.withEnvPrefix(target, 'cluster'),
      containerInsights: true,
      vpc: vpc,
      enableFargateCapacityProviders: true,
    })
    const fargateService = new FargateService(this, 'FargateService', {
      cluster,
      vpcSubnets: vpc.selectSubnets({ subnetType: SubnetType.PRIVATE }),
      taskDefinition: taskDef,
      desiredCount: 1,
    })
  }
}

const discordFunction = (
  scope: cdk.Construct,
  region: string,
  account: string,
  resource: string,
  target: environment.Environments,
  api: RestApi,
) => {
  const requestValidator = fromDiscordRequestValidator(scope, 'validator', api)
  const func = new Function(scope, resource, {
    functionName: `${environment.withEnvPrefix(target, resource)}`,
    code: code(resource),
    handler: 'bin/main',
    timeout: cdk.Duration.minutes(1),
    runtime: Runtime.GO_1_X,
    tracing: Tracing.ACTIVE,
    environment: environmentVariables(target),
  })
  basicPolicytStatements(region, account, target).forEach((s) =>
    func.addToRolePolicy(s),
  )
  api.root.addMethod(
    'POST',
    new LambdaIntegration(func, fromDiscordIntegrationOptions()),
    discordMethodIntegration(requestValidator),
  )
  return func
}

const discordMethodIntegration = (validator: RequestValidator) => {
  return {
    requestValidator: validator,
    apiKeyRequired: false,
    methodResponses: [
      {
        statusCode: '200',
      },
      {
        statusCode: '401',
      },
    ],
  }
}

const fromDiscordRequestValidator = (
  scope: cdk.Construct,
  id: string,
  api: RestApi,
) => {
  return new RequestValidator(scope, id, {
    restApi: api,
    validateRequestBody: true,
    validateRequestParameters: true,
  })
}

const fromDiscordIntegrationOptions = () => {
  return {
    proxy: false,
    requestTemplates: {
      'application/json':
        '{\r\n\
          "timestamp": "$input.params(\'x-signature-timestamp\')",\r\n\
          "signature": "$input.params(\'x-signature-ed25519\')",\r\n\
          "jsonBody" : $input.json(\'$\')\r\n\
        }',
    },
    integrationResponses: [
      {
        statusCode: '200',
      },
      {
        statusCode: '401',
        selectionPattern: '.*[UNAUTHORIZED].*',
        responseTemplates: {
          'application/json': 'invalid request signature',
        },
      },
    ],
  }
}

export const basicPolicytStatements = (
  region: string,
  account: string,
  target: environment.Environments,
) => {
  return [
    dataSourceReadWritePolicyStatement(region, account, target),
    new PolicyStatement({
      effect: Effect.ALLOW,
      actions: ['ssm:Get*'],
      resources: [
        `arn:aws:ssm:${region}:${account}:parameter/claime-verifier*`,
      ],
    }),
  ]
}

const code = (dirname: string) => {
  return Code.fromAsset(
    resolve(`${__dirname}/../`, 'lib', 'functions', dirname, 'bin', 'main.zip'),
  )
}
