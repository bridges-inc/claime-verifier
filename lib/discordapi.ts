import { LambdaIntegration, RestApi } from '@aws-cdk/aws-apigateway'
import { Function, Runtime, Tracing } from '@aws-cdk/aws-lambda'
import * as cdk from '@aws-cdk/core'
import {
  addCorsOptions,
  code,
  environmentVariables,
  withCustomDomain,
} from './api'
import { basicPolicytStatements } from './discord'
import * as environment from './env'

export class DiscordApiStack extends cdk.Stack {
  constructor(
    scope: cdk.Construct,
    id: string,
    target: environment.Environments,
    props?: cdk.StackProps,
  ) {
    super(scope, id, props)
    const api = new RestApi(this, 'RestApi', {
      restApiName: environment.withEnvPrefix(target, 'discord-restapi'),
    })
    const func = discordFunction(
      this,
      this.region,
      this.account,
      'verifydiscord',
      target,
      api,
    )
    withCustomDomain(this, api, restApiDomainName(target), target)
  }
}
// GET /${eoa}?type=domain
// -> いったんbool,at,actual
// -> (ゆくゆくは検証NGだった場合に理由出したい)
const discordFunction = (
  scope: cdk.Construct,
  region: string,
  account: string,
  resource: string,
  target: environment.Environments,
  api: RestApi,
) => {
  const func = new Function(scope, resource, {
    functionName: `${environment.withEnvPrefix(target, resource)}`,
    code: code(resource),
    handler: 'bin/main',
    timeout: cdk.Duration.minutes(1),
    runtime: Runtime.GO_1_X,
    environment: environmentVariables(target),
    tracing: Tracing.ACTIVE,
  })
  basicPolicytStatements(region, account, target).forEach((s) =>
    func.addToRolePolicy(s),
  )
  const rs = api.root.addResource('verify')
  rs.addMethod('POST', new LambdaIntegration(func))
  addCorsOptions(rs, { methods: ['POST'] })
  return func
}

const restApiDomainName = (target: environment.Environments) => {
  return `discord.` + environment.valueOf(target).rootDomain
}
