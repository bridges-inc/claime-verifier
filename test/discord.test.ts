import { SynthUtils } from '@aws-cdk/assert'
import { App } from '@aws-cdk/core'
import { Environments } from '../lib/env'
import { DiscordStack } from './../lib/discord'

test('resource created', () => {
  const app = new App()
  const target = Environments.TEST
  const stack = new DiscordStack(app, 'test', target)
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
})
