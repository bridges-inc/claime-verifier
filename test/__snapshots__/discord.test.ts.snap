// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resource created 1`] = `
Object {
  "Outputs": Object {
    "RestApiEndpoint0551178A": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "RestApi0C43BF4B",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "[HASH REMOVED]",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "[HASH REMOVED]": Object {
      "Description": "[HASH REMOVED]",
      "Type": "String",
    },
    "[HASH REMOVED]": Object {
      "Description": "[HASH REMOVED]",
      "Type": "String",
    },
    "[HASH REMOVED]": Object {
      "Description": "[HASH REMOVED]",
      "Type": "String",
    },
  },
  "Resources": Object {
    "RestApi0C43BF4B": Object {
      "Properties": Object {
        "ApiKeySourceType": "HEADER",
        "Name": "claime-verifier-restapi-test",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "RestApiAccount7C83CF5A": Object {
      "DependsOn": Array [
        "RestApi0C43BF4B",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "RestApiCloudWatchRoleE3ED6605",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
    },
    "RestApiCloudWatchRoleE3ED6605": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "[HASH REMOVED]": Object {
      "DependsOn": Array [
        "RestApiPOSTEF2EB351",
        "validator61463DE6",
      ],
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "RestApi0C43BF4B",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "[HASH REMOVED]": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "[HASH REMOVED]",
        },
        "RestApiId": Object {
          "Ref": "RestApi0C43BF4B",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "RestApiPOSTApiPermissionTesttestRestApi41EB688DPOST7654BF68": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "discordE3A798ED",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "RestApi0C43BF4B",
              },
              "/test-invoke-stage/POST/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "RestApiPOSTApiPermissiontestRestApi41EB688DPOSTC600A146": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "discordE3A798ED",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "RestApi0C43BF4B",
              },
              "/",
              Object {
                "Ref": "[HASH REMOVED]",
              },
              "/POST/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "RestApiPOSTEF2EB351": Object {
      "Properties": Object {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": Object {
          "IntegrationHttpMethod": "POST",
          "IntegrationResponses": Array [
            Object {
              "StatusCode": "200",
            },
            Object {
              "ResponseTemplates": Object {
                "application/json": "invalid request signature",
              },
              "SelectionPattern": ".*[UNAUTHORIZED].*",
              "StatusCode": "401",
            },
          ],
          "RequestTemplates": Object {
            "application/json": "{
          "timestamp": "$input.params('x-signature-timestamp')",
          "signature": "$input.params('x-signature-ed25519')",
          "jsonBody" : $input.json('$')
        }",
          },
          "Type": "AWS",
          "Uri": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                Object {
                  "Fn::GetAtt": Array [
                    "discordE3A798ED",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "MethodResponses": Array [
          Object {
            "StatusCode": "200",
          },
          Object {
            "StatusCode": "401",
          },
        ],
        "RequestValidatorId": Object {
          "Ref": "validator61463DE6",
        },
        "ResourceId": Object {
          "Fn::GetAtt": Array [
            "RestApi0C43BF4B",
            "RootResourceId",
          ],
        },
        "RestApiId": Object {
          "Ref": "RestApi0C43BF4B",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "claimeverifierdiscordapikeytest75609FE8": Object {
      "Properties": Object {
        "GenerateSecretString": Object {},
      },
      "Type": "AWS::SecretsManager::Secret",
    },
    "discordE3A798ED": Object {
      "DependsOn": Array [
        "discordServiceRoleDefaultPolicyE970AC2D",
        "discordServiceRole5137BD59",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "[HASH REMOVED]",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "[HASH REMOVED]",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "[HASH REMOVED]",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "FunctionName": "claime-verifier-discord-test",
        "Handler": "bin/main",
        "Role": Object {
          "Fn::GetAtt": Array [
            "discordServiceRole5137BD59",
            "Arn",
          ],
        },
        "Runtime": "go1.x",
        "Timeout": 60,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "discordServiceRole5137BD59": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "discordServiceRoleDefaultPolicyE970AC2D": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "dynamodb:Put*",
                "dynamodb:Get*",
                "dynamodb:Scan*",
                "dynamodb:Delete*",
                "dynamodb:Batch*",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:dynamodb:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":table/claime-verifier-main*",
                  ],
                ],
              },
            },
            Object {
              "Action": "ssm:Get*",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/claime-verifier*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "secretsmanager:GetSecretValue",
                "secretsmanager:DescribeSecret",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "claimeverifierdiscordapikeytest75609FE8",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "discordServiceRoleDefaultPolicyE970AC2D",
        "Roles": Array [
          Object {
            "Ref": "discordServiceRole5137BD59",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "validator61463DE6": Object {
      "Properties": Object {
        "RestApiId": Object {
          "Ref": "RestApi0C43BF4B",
        },
        "ValidateRequestBody": true,
        "ValidateRequestParameters": true,
      },
      "Type": "AWS::ApiGateway::RequestValidator",
    },
  },
}
`;
